modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/ClusterManagementGroupManager.java

```diff
@@ -424,10 +434,34 @@ private CompletableFuture<Void> updateLogicalTopology(CmgRaftService service) {
                 }));
     }
 
-    private void handleCancelInit(CancelInitMessage msg) {
+    private void handleCancelInit(CancelInitMessage msg, Consumer<Throwable> onError) {
         LOG.info("CMG initialization cancelled [reason={}]", msg.reason());
+        this.scheduledExecutor.execute(() -> this.destroyCmgWithEvents().exceptionally(toFunction(onError)));
+    }
+
+    /** Converts the consumer to a function that returns null. */
+    private static <T> Function<T, Void> toFunction(Consumer<T> consumer) {
+        return arg -> {
+            consumer.accept(arg);
+            return null;
+        };
+    }
+
+    /** Delegates call to {@link #destroyCmg()} but fires the associated events. */
+    private CompletableFuture<Void> destroyCmgWithEvents() {
+        LOG.info("CMG cancellation procedure started");
+        return inBusyLockAsync(busyLock,
+                () -> fireEvent(ClusterManagerGroupEvent.BEFORE_DESTROY_RAFT_GROUP, EmptyEventParameters.INSTANCE)
+                    .handle((none, err) -> {
```

Comment:
we should use `handleAsync` here, just in case

Reply:
Hey @sashapolo sorry I didn't understand your idea.
Do you mean to use `handleAsync` with the default java threadpool or with `this.scheduledExecutor` just to bump the callback to the back of the executor's queue??

Reply:
I mean the latter, we try not to use the default thread pool in the server code. The idea is that event listeners return futures which may be executed on pools, unknown to the current component, so it's better to re-dispatch the `handle` method to be sure that we will not block listener's threads on accident.